hathor/event/websocket/protocol.py

```diff
@@ -0,0 +1,186 @@
+            raise InvalidRequestError(InvalidRequestType.STREAM_IS_ACTIVE)
+
+        self._validate_ack(request.last_ack_event_id)
+
+        self._last_sent_event_id = request.last_ack_event_id
+        self._ack_event_id = request.last_ack_event_id
+        self._window_size = request.window_size
+        self._stream_is_active = True
+
+        self.factory.send_next_event_to_connection(self)
+
+    def _handle_ack_request(self, request: AckRequest) -> None:
+        if not self._stream_is_active:
+            raise InvalidRequestError(InvalidRequestType.STREAM_IS_INACTIVE)
+
+        self._validate_ack(request.ack_event_id)
+
+        self._ack_event_id = request.ack_event_id
+        self._window_size = request.window_size
+
+        self.factory.send_next_event_to_connection(self)
+
+    def _handle_stop_stream_request(self) -> None:
+        if not self._stream_is_active:
+            raise InvalidRequestError(InvalidRequestType.STREAM_IS_INACTIVE)
+
+        self._stream_is_active = False
+
+    def _validate_ack(self, ack_event_id: Optional[int]) -> None:
+        """Validates an ack_event_id from a request.
+
+        The ack_event_id can't be smaller than the last ack we've received
+        and can't be larger than the last event we've sent.
+        """
+        if self._ack_event_id is not None and (
+            ack_event_id is None or ack_event_id < self._ack_event_id
+        ):
+            raise InvalidRequestError(InvalidRequestType.ACK_TOO_SMALL)
+
+        if ack_event_id is not None and (
+            self._last_sent_event_id is None or self._last_sent_event_id < ack_event_id
+        ):
+            raise InvalidRequestError(InvalidRequestType.ACK_TOO_LARGE)
+
+    def send_event_response(self, event_response: EventResponse) -> None:
+        self._send_response(event_response)
+        self._last_sent_event_id = event_response.event.id
+
+    def send_invalid_request_response(
+        self,
+        _type: InvalidRequestType,
+        invalid_payload: Optional[bytes] = None,
+        error_message: Optional[str] = None
+    ) -> None:
+        invalid_request = None if invalid_payload is None else invalid_payload.decode('utf8')
+        response = InvalidRequestResponse(
+            type=_type,
+            invalid_request=invalid_request,
+            error_message=error_message
+        )
+
+        self._send_response(response)
+
+    def _send_response(self, response: Response) -> None:
+        payload = json_dumpb(response.dict())
+
+        try:
+            self.sendMessage(payload)
+        except Disconnected:
+            # Connection is closed. Nothing to do.
+            pass
+        # XXX: unfortunately autobahn can raise 3 different exceptions and one of them is a bare Exception
+        # https://github.com/crossbario/autobahn-python/blob/v20.12.3/autobahn/websocket/protocol.py#L2201-L2294
+        except Exception:
+            self.log.error('send failed, moving on', exc_info=True)
```

Comment:
Should we close the connection?

Reply:
I'm not sure. I copied this handling pattern from [factory.py#L218:L226](https://github.com/HathorNetwork/hathor-core/blob/85206cb631b609a5680e276e4db8cffbb418eb88/hathor/websocket/factory.py#L218:L226).

From my understanding of autobahn's code, they raise a raw `Exception` when there's an invalid payload fragment size, which we don't use, so in theory we'll never encounter this error. If it happens, I don't know if closing the connection and then reopening it would make the error go away. Maybe it would just keep raising it. I would say it's as uncertain as just trying to send the message again (that's the current behavior).

Does @jansegre have any opinion on this?

Reply:
Got it. I feel it's ok to move forward the way it is. Let's see if @jansegre has something to add.

Reply:
I'm not too sure either. I lean more to closing the connection, even if it's different from the other websocket.

Reply:
But I'm OK moving forward as is.

Reply:
Ok, then I'll keep it like this, for now. We'll keep an eye and eventually we could open another PR to improve this and other websockets in the project.